### PR TITLE
feat: tray & app menu item "Check for Updates..."

### PR DESCRIPTION
**Pull request by bajtos**
Add "Check for Updates..." to the Tray menu. On macOS, add this item to the Application menu too. While the check is running, replace "Check for Updates" with a disabled item "Checking for Updates".

See https://github.com/filecoin-project/filecoin-station/issues/53

The implementation is heavily inspired by IPFS Desktop:
- [src/auto-updater/index.js](https://github.com/ipfs/ipfs-desktop/blob/b269cde5231915d2930eefba55b706012843a155/src/auto-updater/index.js)
- [src/tray.js](https://github.com/ipfs/ipfs-desktop/blob/b269cde5231915d2930eefba55b706012843a155/src/tray.js)

### Out of scope

- Reworking the OS-level notification
- Showing a banner at the top of the web UI when an update is ready

### Screenshots

A new macOS application menu item:

Screenshot 2022-07-27 at 14 48 56

While checking for updates:

Screenshot 2022-07-27 at 14 49 06

A new Tray menu item:

Screenshot 2022-07-27 at 14 50 35

While checking for updates:

Screenshot 2022-07-27 at 14 50 41

Various dialogs shown by the app:

Screenshot 2022-07-27 at 14 49 33

Screenshot 2022-07-27 at 14 51 43
Screenshot 2022-07-27 at 14 51 56
Screenshot 2022-07-27 at 14 52 05

The annoying notification we should remove later:

Screenshot 2022-07-27 at 14 51 24

### CODE REVIEW

File: main/app-menu.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    id: 'checkForUpdates',
+    label: 'Check For Updates...',
+    click: () => { ctx.manualCheckForUpdates() }
+  }))
+  menu.items[0].submenu?.insert(2, new MenuItem({
+    id: 'checkingForUpdates',
+    label: 'Checking For Updates',
+    enabled: false,
+    visible: false
+  }))
```
Comment (bajtos) [inline main/app-menu.js:21]: I am not sure where other Electron applications place "Check for Updates..." item on Windows & Linux, so I am adding this feature to macOS only to keep things simple.

If you think we should add this item to the application menu on other platforms too, then let's create follow-up GH issues, one per platform.

File: main/updater.js [hunk 2]
```diff
@@ -1,69 +1,158 @@
...
+    checkingManually = true
+    checkForUpdatesInBackground()
+  }
+}
+
+function checkForUpdatesInBackground () {
+  ipcMain.emit(ipcMainEvents.UPDATE_CHECK_STARTED)
+
+  // TODO: replace this with autoUpdater.checkForUpdatesAndNotify()
+  autoUpdater.checkForUpdates()
```
Comment (bajtos) [inline main/updater.js:60]: This todo comment is coming from the original source code. I am not sure if it's still relevant. Do you remember @lidel why you would like to switch from `autoUpdater.checkForUpdates()` to `autoUpdater.checkForUpdatesAndNotify()`?

File: main/app-menu.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  Menu.setApplicationMenu(menu)
+  setupIpcEventListeners(menu)
+}
+
+/**
+ * @param {Electron.Menu} appMenu
+ */
+function setupIpcEventListeners (appMenu) {
+  ipcMain.on(ipcMainEvents.UPDATE_CHECK_STARTED, () => {
+    getItemById('checkForUpdates').enabled = false
```
Comment (juliangruber) [inline main/app-menu.js:32]:
```suggestion
```

I don't see the purpose of this, afaic if an item is invisible it's also disabled

File: main/app-menu.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   * @param {string} id
+   * @returns {Electron.MenuItem}
+   */
+  function getItemById (id) {
+    const item = appMenu.getMenuItemById(id)
+    if (!item) throw new Error(`Unknown app menu item id: ${id}`)
+    return item
+  }
```
Comment (juliangruber) [inline main/app-menu.js:51]: Could we also do without this function? We'd get an error in the consuming lines anyway.

File: main/app-menu.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
+const { Menu, MenuItem, ipcMain } = require('electron')
+const { ipcMainEvents } = require('./ipc')
+
+function setupAppMenu (/** @type {import('./typings').Context} */ ctx) {
+  if (process.platform !== 'darwin') return
+
+  // Add "Check for updates..." item to the Application menu on MacOS
```
Comment (juliangruber) [inline main/app-menu.js:7]: Did you consider Windows for this PR as well? If we don't add it now then an issue will help keeping the implementations in sync in the future

File: main/app-menu.js (hunk 3 above)
Comment (juliangruber) [reply main/app-menu.js:32]: Same question applies to the tray configuration

File: main/app-menu.js (hunk 4 above)
Comment (juliangruber) [reply main/app-menu.js:51]: Same question applies to the tray configuration

File: main/updater.js [hunk 6]
```diff
@@ -1,69 +1,158 @@
...
+        buttons: ['Close']
+      })
     }
+    return
+  }
+
+  setup(ctx)
+
+  checkForUpdatesInBackground() // async check on startup
+  setInterval(checkForUpdatesInBackground, 43200000) // every 12 hours
```
Comment (juliangruber) [inline main/updater.js:47]:
```suggestion
  setInterval(checkForUpdatesInBackground, ms('12h'))
```

https://github.com/vercel/ms helps clean up some comments

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: main/app-menu.js (hunk 3 above)
Comment (bajtos) [reply main/app-menu.js:32]: To be honest, I am not sure about this either. I have simply copied what I found in ipfs-desktop:

https://github.com/ipfs/ipfs-desktop/blob/b269cde5231915d2930eefba55b706012843a155/src/tray.js#L305-L307

```js
    menu.getMenuItemById('checkForUpdates').enabled = !isUpdating
    menu.getMenuItemById('checkForUpdates').visible = !isUpdating
    menu.getMenuItemById('checkingForUpdates').visible = isUpdating
```

File: main/updater.js (hunk 6 above)
Comment (bajtos) [reply main/updater.js:47]: Personally, I am cautious about adding new dependencies to the project. Is the maintenance cost of keeping `ms` up to date worth the benefit of avoiding a comment?

BTW my PR is not changing this part, I just moved existing code & the comment.

File: main/app-menu.js (hunk 5 above)
Comment (bajtos) [reply main/app-menu.js:7]: We are thinking alike 😀

https://github.com/filecoin-project/filecoin-station/pull/62#discussion_r931039942

File: main/app-menu.js (hunk 1 above)
Comment (juliangruber) [reply main/app-menu.js:21]: I think we should add these items there too 👍 

https://github.com/filecoin-project/filecoin-station/issues/63
https://github.com/filecoin-project/filecoin-station/issues/64

File: main/updater.js (hunk 6 above)
Comment (juliangruber) [reply main/updater.js:47]: To me this is a DX improvement that overweighs the maintenance aspect, especially once we use it in more places. In my experience of using `ms` it hasn't been a maintenance burden.

File: main/app-menu.js (hunk 1 above)
Comment (bajtos) [reply main/app-menu.js:21]: Thanks for creating the GH issues, @juliangruber 🙇🏻 I'll mention them in a code comment.

File: main/app-menu.js (hunk 4 above)
Comment (bajtos) [reply main/app-menu.js:51]: This function serves two purposes:
- Make the TypeScript compiler happy
- Report a more helpful error message than _Uncaught TypeError: Cannot set properties of undefined (setting 'visible')`_

**Comment by bajtos** [comment]
@juliangruber I addressed your feedback in [172bf3b](https://github.com/filecoin-project/filecoin-station/pull/62/commits/172bf3bd193196a1d203c7d20a475bc5b783a509).

- Added a code comment about Windows & Linux
- Reworked update check interval `ms('12h')` 

LGTY now?

File: main/app-menu.js (hunk 4 above)
Comment (juliangruber) [reply main/app-menu.js:52]: Is this even run through ts, given the .js extension?

File: main/app-menu.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    getItemById('checkForUpdates').visible = false
+    getItemById('checkingForUpdates').visible = true
+  })
+
+  ipcMain.on(ipcMainEvents.UPDATE_CHECK_FINISHED, () => {
+    getItemById('checkForUpdates').visible = true
+    getItemById('checkingForUpdates').visible = false
+  })
+
+  /**
```
Comment (juliangruber) [inline main/app-menu.js:44]:
```suggestion
  /**
   * Get item or fail with a useful error message.
```

File: main/app-menu.js (hunk 4 above)
Comment (bajtos) [reply main/app-menu.js:52]: Yes, we check types in JS files, that's why there are JSDoc comments in most places.

https://github.com/filecoin-project/filecoin-station/blob/62b5de84631c0bd3b01aa70407f4f1ccdf6ca021/tsconfig.json#L7-L8

I found this setup useful so far. I have full auto-completion support in VS Code and the type checker can catch many programming errors.

File: main/app-menu.js (hunk 7 above)
Comment (bajtos) [reply main/app-menu.js:44]: Done in [af82612](https://github.com/filecoin-project/filecoin-station/pull/62/commits/af8261296e09e04876d64135d36588b2360d2e52)

**Review by juliangruber — APPROVED** [review]

File: main/updater.js (hunk 2 above)
Comment (lidel) [reply main/updater.js:62]: I vaguely remember that the rationale was that it is less code to maintain, and that meant smaller surface for breaking autoupdates when upstream APIs change in subtle ways (happened in the past)

**Pull request merged**